### ISSUE TITLE
Could tabs be preserved inside <pre> blocks?

### ISSUE THREAD

**Issue by jacksonp** [htacg/tidy-html5#403]
Running tidy on this doc (→ represents a tab character):

``` html
<!DOCTYPE html>
<title>pre test</title>
<pre>→A,
→B,
→C,</pre>
```

changes how the document looks in browsers, as tidy replaces the tabs with spaces.

Before tidy:

```
    A,
    B,
    C,
```

After tidy:

```
   A,
        B,
        C,
```

**Comment by geoffmcl** [comment]
@jacksonp thanks for the report... actually I think this has been posted before, but can not find it at the moment... maybe only in my imagination...

I think the problem is that tidy outputs -

````html
<pre>   A,
        B,
        C,</pre>
````

That is it reduces the spaces output for the tab by the length of the `<div>` on the same line...

To preserve the view in a browser, tidy should output -

````html
<pre>        A,
        B,
        C,</pre>
````

This needs to be checked, and fixed... any takers?

**Comment by geoffmcl** [comment]
As we are about to `release` **5.4**, but this does remain a sad `Pretty Print` alignment issue, and it is a case where the output of tidy does not keep the browser rendering of the original, am moving it to `next` **5.5**...

I am sure there could be a relatively easy **fix** for this alignment... 

As always, comments, patches, PR very **welcome**... thanks...

**Comment by fyiman** [comment]
One way to tackle it would be to treat `<pre>` as having 0-length.

**Comment by balthisar** [comment]
I'm willing to attempt to tackle this, but what's the approach we want to take? There are two routes here:

- @jacksonp indicates in the issue title that he would preserve tabs in **pre** blocks, whereas Tidy changes them to spaces. The question is, is replacing tabs with spaces in a **pre** block correct/expected behavior? I think there's a strong argument for leaving things that the user indicates are preformatted alone.

- On the other hand, possibly the user expects that tabs will be converted to spaces, and maybe this is why he is using Tidy?

**Comment by geoffmcl** [comment]
@balthisar I certainly agree changing tidy to **not** replace `<pre>` tabs with spaces at this late stage might bring out **all** the users who now **expect** this behaviour, rightly or wrongly. It was that way back in the earliest 2000 version I have... that's more than 17 years of history, expectations, acceptance...

And even then, the first tab in the `<pre>` before the `A` got given less spaces than the `B` or `C`, so this pretty printing `<pre>` spacing bug on the first line of text has been around for a **long** time...

On the other hand, I agree there seems a **strong** argument to leave things alone when `preformatted` indicated... but I guess at this stage that would have to be a new opt-in option...  like say  `--keep-tabs-in-pre yes`, or something...

But it also seems it would not be too difficult to treat the `<pre>` tag as having zero length, as @fyiman suggested, thus produce a spaced out text that matched what is visually expected in a browser... an important criteria...

And it is interesting that `-i --indent-with-tabs yes` does **not** effect this situation. If a `<p>para</p>` is added in front of the `<pre>`, the `<p>`, and the `<pre>` tags will be indented with a **tab**, but the tabs in the text in the `<pre>` will still get converted to spaces...

Have not checked fully, but this new `indent-with-tabs` option may produce other anomolies here... but maybe not...

Personally, I would go for fixing the small `space` problem... that should not be difficult to discount the `<pre>` tag as representing space, and leave the bigger, perhaps different, question of whether there should be a new option to handle `preformatted` text differently to later, separately, but that's just me...

**BUT** as I start to look **deeper** into this, I am starting to feel **both** ideas are much more involved than just a simple rational choice...

First I started to look at the upper level, pretty print output, but as you will see that is not it...

The `<pre>` output is handled in the `Ln:2255:    else /* some kind of container element */`, which also handles the `<textarea>` tag. Before outputting the `<pre>` the internal print buffer has been flushed, so is empty. 

Then `PPrintTag( doc, mode, indent, node );` is done... and then the node content, with `(mode | PREFORMATTED | NOWRAP)` added... which be done in `PPrintText(...)`...

But when I look at the `doc->lexer->lexbuf`, I can see the **problem** is already **there**... the 3 lines of text, spearated by `0x0a` has already been added as -

```hex
20 20 20 41 2c 0a 20 20 20 20 20 20 20 20 42 2c 0a 20 20 20 20 20 20 20 20 43 2c 0a
```

So that means going back down to the text collection from the stream... that is where the `<pre>` caused the first line tab to only add 3 spaces, while the 2nd and 3rd line tabs got converted to 8 spaces each...

So in a way this turns out to **not** be a pretty print problem, but a very low level text collect from the stream problem... maybe not as easy as I first thought... but one never nows until one looks at exactly where it happens... down at `DocParseStream`...

Now when a tab is fetched from the stream, `if (c == '\t')`, the problem is that the `in->curcol` size is used to determine how many spaces to add - the code -

```code
        if (c == '\t')
        {
#ifdef TIDY_STORE_ORIGINAL_TEXT
            added = yes;
            TY_(AddCharToOriginalText)(in, (tchar)c);
#endif
            in->tabs = tabsize > 0 ?
                tabsize - ((in->curcol - 1) % tabsize) - 1
                : 0;
            in->curcol++;
            c = ' ';
            break;
        }
```

And of course, due to the `<pre>`, the current `in->curcol` is 6, thus `in->tabs` get set to 2 - actually 3 minus one space that is being returned now... so the **problem** has been created `streamio.c`...

Eventually we get to the **newline**, `if (c == '\n')`, and `in->curcol` is returned to 1... and on the next tab, `in->tabs` gets set to 7 - actually tabsize, 8, minus one space returned now...

Wow, so it seems quite difficult to inform this very low level io service that in this case `in->curcol` needs to be ignored... can something be done with `in->lastcols[LASTPOS_SIZE]`? ... We are keeping the stream io position... could we also keep a document column position, that would know a html tag takes no column size? or something...

Somehow, I now do **not** see this as a simple fix!

And given that I now see, remind myself, this low stream io does the tab to space conversion, even an idea to **not** convert tabs, is some case, would also not be trivial... It would mean tabs would have to be added to the lexer buffer, and the choice of tab-to-space conversion done later, probably during output...

This needs a **lot** more investigation...

But HTH!

OT: This will probably be my last tidy comment/action before getting back in mid Sept... and really, **really** thanks for stepping in, especially doing the merges... many thanks... g.

**Comment by geoffmcl** [comment]
Have pushed a test branch, `keep-tabs`, that has a new option, `--keep-tabs yes`... it defaults `no`.

At this stage have only done some preliminary tests. It certainly **fixes** this specific issue sample, but have yet to do more extensive tests...

The full diff is given below to facilitate a quick review...

```diff
diff --git a/include/tidyenum.h b/include/tidyenum.h
index 90d7352..ae523a6 100644
--- a/include/tidyenum.h
+++ b/include/tidyenum.h
@@ -604,6 +604,7 @@ typedef enum
     TidyJoinClasses,             /**< Join multiple class attributes */
     TidyJoinStyles,              /**< Join multiple style attributes */
     TidyKeepFileTimes,           /**< If yes last modied time is preserved */
+    TidyKeepTabs,                /**< If yes keep input source tabs */
     TidyLiteralAttribs,          /**< If true attributes may use newlines */
     TidyLogicalEmphasis,         /**< Replace i by em and b by strong */
     TidyLowerLiterals,           /**< Folds known attribute values to lower case */
diff --git a/src/config.c b/src/config.c
index 2fd89d7..7b7bd2e 100644
--- a/src/config.c
+++ b/src/config.c
@@ -214,6 +214,7 @@ static const TidyOptionImpl option_defs[] =
     { TidyJoinClasses,             MX, "join-classes",                BL, no,              ParsePickList,     &boolPicks          },
     { TidyJoinStyles,              MX, "join-styles",                 BL, yes,             ParsePickList,     &boolPicks          },
     { TidyKeepFileTimes,           IO, "keep-time",                   BL, no,              ParsePickList,     &boolPicks          },
+    { TidyKeepTabs,                PP, "keep-tabs",                   BL, no,              ParsePickList,     &boolPicks          }, /* 20171103 - Issue #403 */
     { TidyLiteralAttribs,          MR, "literal-attributes",          BL, no,              ParsePickList,     &boolPicks          },
     { TidyLogicalEmphasis,         MC, "logical-emphasis",            BL, no,              ParsePickList,     &boolPicks          },
     { TidyLowerLiterals,           MR, "lower-literals",              BL, yes,             ParsePickList,     &boolPicks          },
diff --git a/src/language_en.h b/src/language_en.h
index f56f1ec..9f5b93e 100644
--- a/src/language_en.h
+++ b/src/language_en.h
@@ -700,6 +700,17 @@ static languageDefinition language_en = { whichPluralForm_en, {
         "Note this feature is not supported on some platforms. "
     },
     {/* Important notes for translators:
+     - Use only <code></code>, <var></var>, <em></em>, <strong></strong>, and
+     <br/>.
+     - Entities, tags, attributes, etc., should be enclosed in <code></code>.
+     - Option values should be enclosed in <var></var>.
+     - It's very important that <br/> be self-closing!
+     - The strings "Tidy" and "HTML Tidy" are the program name and must not
+     be translated. */
+        TidyKeepTabs,            0,
+        "This option specifies if Tidy should keep tabs found in the source. "
+    },
+    {/* Important notes for translators:
       - Use only <code></code>, <var></var>, <em></em>, <strong></strong>, and
       <br/>.
       - Entities, tags, attributes, etc., should be enclosed in <code></code>.
diff --git a/src/streamio.c b/src/streamio.c
index cd36efe..bf75a1a 100644
--- a/src/streamio.c
+++ b/src/streamio.c
@@ -239,7 +239,6 @@ static void RestoreLastPos( StreamIn *in )
 uint TY_(ReadChar)( StreamIn *in )
 {
     uint c = EndOfStream;
-    uint tabsize = cfg( in->doc, TidyTabSize );
 
     if ( in->pushed )
         return PopChar( in );
@@ -269,11 +268,15 @@ uint TY_(ReadChar)( StreamIn *in )
 
         if (c == '\t')
         {
-            in->tabs = tabsize > 0 ?
-                tabsize - ((in->curcol - 1) % tabsize) - 1
-                : 0;
+            Bool keeptabs = cfg( in->doc, TidyKeepTabs );
             in->curcol++;
-            c = ' ';
+            if (!keeptabs) {
+                uint tabsize = cfg(in->doc, TidyTabSize);
+                in->tabs = tabsize > 0 ?
+                    tabsize - ((in->curcol - 1) % tabsize) - 1
+                    : 0;
+                c = ' ';
+            }
             break;
         }
 
```

I have put it in the `PP` category, **TidyPrettyPrint**, but since this happens at the IO level, maybe that should be `IO`, **TidyFileIO**, or some other category. Also maybe the simple text explanation could be expanded to include information on tidy's default IO level tab replacement...

And maybe this is too general since it will effect `tabs` **anywhere** in the html code, **not** just in a `<pre>` block, and maybe that could have other consequences...

Should have no effect on the `regression` tests, should there be any tabs there, since it defaults to `no`...

Would be appreciated if others could pull and test this branch, and provide feedback... thanks...

**Comment by geoffmcl** [comment]
Created PR #641 to address this, after full testing and agreement... thanks...

**Comment by balthisar** [comment]
@geoffmcl, on quick testing, it certainly seems to work! On quick testing, though, it doesn't seem to do _exactly_ what you describe (or what's written in the new description). I think it _does_ address this feature request, though.

I see what you're doing in `streamio.c`, but somewhere, somehow (during lexing? during parsing? during pretty printing?) the tabs _are_ being replaced with spaces, _except_ in **pre** tags, and so I think the behavior is correct, just not the description.

For example, if I `--indent yes --indent-spaces 3 --keeptabs yes`, and my source file is tab indented, it will (properly) indent with spaces, replacing the source tabs. Tabs in, say, **li** elements are replaced with spaces. Multiple subsequent tabs are coalesced to a single space. Attribute values with tabs are changed to spaces (and coalesced if multiple, but `literal-attributes` still works properly).

So, basically, it's working as @jacksonp requested, and _not_ as described, which I think is actually the correct behavior.

I support the PR, with an improved description, and maybe an understanding of why it works even though there should be tabs everywhere (I'll dig into it if I have time; probably pretty print is making sure we don't output tabs).

**Comment by geoffmcl** [comment]
@balthisar thanks for testing...

Yes, now in more testing, adding tabs **everywhere**, I too see they are replaced with **a** space, seemingly in nearly everything but a `<pre>`, as this issue requests...

Not only should the description be changed, but perhaps the option should be renamed `keep-pre-tabs` to make it real clear...

It seems the `lexer` guards against adding tabs with -

```C
                if (TY_(IsWhite)(c))
                {
                    /* was previous character white? */
                    if (lexer->waswhite)
                    {
                        if (mode != Preformatted && mode != IgnoreMarkup)
                        {
                            --(lexer->lexsize);
                            SetLexerLocus( doc, lexer );
                        }
                    }
                    else /* prev character wasn't white */
                    {
                        lexer->waswhite = yes;

                        if (mode != Preformatted && mode != IgnoreMarkup && c != ' ')
                            ChangeChar(lexer, ' ');
                    }

                    continue;
                }
```

Note it not only changes non-space `white` chars to a space, but dumps them if more than one... and of course when the `<pre>` is being collected, the mode is `Preformatted` so the tabs are left alone, if the option is `yes` and they are fed from the stream...

But that leaves another case when they will be kept, and that is when the mode is `IgnoreMarkup`... When does that happen? Maybe it is more than just the `<pre>`... 

Need to test, and understand more... **more** feedback very welcome... thanks...

**Comment by geoffmcl** [comment]
Ok, seems tabs are also retained in a `<script>` block... so can not use `keep-pre-tabs` as the option name... and further alters any description of the option...

Also thought they might be retained in a `<style>` block, but seems not...

Still testing... but as stated **more** feedback very welcome... thanks...

**Comment by geoffmcl** [comment]
This is getting **crazy** to exactly tie down which source `tabs` will be **kept**, and which **converted** to spaces, and under what **circumstances**...

Now it seems **some** will be kept in a `<style>` block if indent, `-i`, is **not** used!?!? How can that be? Or more importantly, how can that be easily described?

Just pushed yet another sample [in_403-2.html](https://github.com/geoffmcl/tidy-test/blob/master/test/input5/in_403-2.html) to my test repo, which demonstrates this...

But what else is effected? Not easy to logically decide without actual trying a sample...

Still testing, but **need** help... thanks...

**Comment by geoffmcl** [comment]
After **lots** more testing, have come up with the following quite **full** description of this **new** `keep-tabs` option -

```
        "With the default <var>no</var> Tidy will replace all source tabs, with spaces, "
        "controlled by the option <code>tab-size</code>, and the current line offset. "
        "<br/>"
        "If set <var>yes</var> this option specifies Tidy should keep certain tabs "
        "found in the source, but only "
        "in preformatted blocks like <code>&lt;pre&gt;</code>, and other CDATA elements like "
        "<code>&lt;script&gt;</code>, <code>&lt;style&gt;<code>, and other pseudo elements like "
        "<code>&lt;?php ... ?&gt;</code>. As always, all other tabs, or sequences of tabs, in "
        "the source will continue to be replaced with a space. "
```

Comments sought before pushing this to PR #641... thanks...

Have also pushed a small special debug change to be able to `see` tabs in text...

**Comment by balthisar** [comment]
@geoffmcl, by "current line offset," do you mean indentation level?

In any case, it looks good.

**Comment by geoffmcl** [comment]
@balthisar Thanks for the review... err, no - it is this "current line offset," that causes this bug in the first place...

Given source lines of `<pre>\tA\n\tB`, the first tab will be expanded to 2/3 spaces, while the second to 7/8 spaces, thus the **visual** rendering bug in a browser...

Have expanded the description a little more -

```diff
diff --git a/src/language_en.h b/src/language_en.h
index 9f5b93e..56b91e8 100644
--- a/src/language_en.h
+++ b/src/language_en.h
@@ -708,7 +708,17 @@ static languageDefinition language_en = { whichPluralForm_en, {
      - The strings "Tidy" and "HTML Tidy" are the program name and must not
      be translated. */
         TidyKeepTabs,            0,
-        "This option specifies if Tidy should keep tabs found in the source. "
+        "With the default <var>no</var> Tidy will replace all source tabs, with spaces, "
+        "controlled by the option <code>tab-size</code>, and the current line offset. "
+        "Of course, except in the special blocks/elments enumerated below, this will later "
+        "be reduced to just one space. "
+        "<br/>"
+        "If set <var>yes</var> this option specifies Tidy should keep certain tabs "
+        "found in the source, but only "
+        "in preformatted blocks like <code>&lt;pre&gt;</code>, and other CDATA elements like "
+        "<code>&lt;script&gt;</code>, <code>&lt;style&gt;<code>, and other pseudo elements like "
+        "<code>&lt;?php ... ?&gt;</code>. As always, all other tabs, or sequences of tabs, in "
+        "the source will continue to be replaced with a space. "
     },
     {/* Important notes for translators:
       - Use only <code></code>, <var></var>, <em></em>, <strong></strong>, and
```

Have now pushed this change to PR #641, which I will merge shortly, barring any negative feedback... thanks...

**Comment by balthisar** [comment]
This is merged and alive! Thanks for all the hard work.

**Comment by geoffmcl** [comment]
Seems 5.5.85 passes all regression tests, so **now** this could be closed... 

Oops, someone already did that **prematurely** ;=))
